src/dirent_lookup.h

```diff
@@ -187,53 +189,57 @@ typename DirentLookup<TConfig>::Result
 FastDirentLookup<TConfig>::find(char ns, const std::string& key) const
 {
   const auto r = lookupGrid.getRange(ns + key);
-  entry_index_type l(r.begin);
-  entry_index_type u(r.end);
-
-  if (l == u)
-    return {false, entry_index_t(l)};
-
-  return BaseType::findInRange(l, u, ns, key);
+  return BaseType::findInRange(r.begin, r.end, ns, key);
 }
 
 template<typename TConfig>
 typename DirentLookup<TConfig>::Result
 DirentLookup<TConfig>::find(char ns, const std::string& key) const
 {
-  if ( direntCount == 0 )
-      return { false, entry_index_t(0) };
+  return findInRange(0, direntCount, ns, key);
+}
+
+template<typename TConfig>
+typename DirentLookup<TConfig>::Result
+DirentLookup<TConfig>::findInRange(entry_index_type l, entry_index_type u, char ns, const std::string& key) const
+{
+  if ( l == u )
+      return { false, entry_index_t(l) };
 
-  const auto c = compareWithDirentAt(ns, key, 0);
+  const auto c = compareWithDirentAt(ns, key, l);
   if ( c < 0 )
-      return { false, entry_index_t(0) };
+      return { false, entry_index_t(l) };
   else if ( c == 0 )
-      return { true, entry_index_t(0) };
+      return { true, entry_index_t(l) };
 
-  if ( compareWithDirentAt(ns, key, direntCount-1) > 0 )
-      return { false, entry_index_t(direntCount) };
+  if ( compareWithDirentAt(ns, key, u-1) > 0 )
+      return { false, entry_index_t(u) };
 
-  return findInRange(0, direntCount, ns, key);
+  return binarySearchInRange(l, u-1, ns, key);
```

Comment:
Shouldn't be `binarySearchInRange(l, u, ns, key);` ?

Reply:
The entry at `u-1` has already been checked above. Execution reaches here if `ns/key <= entry[u-1]`. Therefore `u-1` is a valid upper bound for our search range.

Reply:
You're right. This is more that you have changed the behavior of `binarySearchInRange` (upper bound `>=`) compared to `findInRange` (upper bound `>`)